File: sirius-core/src/main/java/org/starcoin/sirius/core/AccountInformation.java
package org.starcoin.sirius.core;

import org.starcoin.core.BlockAddress;
import org.starcoin.core.Hash;
import org.starcoin.core.ProtobufCodec;
import org.starcoin.proto.Starcoin;
import org.starcoin.proto.Starcoin.ProtoAccountInfo;

import java.util.Objects;

public class AccountInformation implements ProtobufCodec<Starcoin.ProtoAccountInfo> {

    public static final AccountInformation EMPTY_ACCOUNT =
            new AccountInformation(BlockAddress.DEFAULT_ADDRESS, 0, null);

    // just keep hash of address
    private Hash address;
    private long allotment;
    private Update update;

    public AccountInformation() {
    }

    public AccountInformation(BlockAddress address, long allotment, Update update) {
        this.address = Hash.of(address.toBytes());
        this.allotment = allotment;
        this.update = update;
    }

    public AccountInformation(Hash address, long allotment, Update update) {
        this.address = address;
        this.allotment = allotment;
        this.update = update;
    }

    public Hash getAddress() {
        return address;
    }

    public Update getUpdate() {
        return update;
    }

    public void setUpdate(Update update) {
        this.update = update;
    }

    public long getAllotment() {
        return allotment;
    }

    @Override
    public Starcoin.ProtoAccountInfo marshalProto() {
        ProtoAccountInfo.Builder builder =
                Starcoin.ProtoAccountInfo.newBuilder()
                        .setAddress(this.getAddress().toProto())
                        .setAllotment(this.getAllotment());
        if (this.update != null) {
            builder.setUpdate(this.getUpdate().toProto());
        }
        return builder.build();
    }

    @Override
    public void unmarshalProto(Starcoin.ProtoAccountInfo proto) {
        this.address = Hash.wrap(proto.getAddress());
        this.allotment = proto.getAllotment();
        this.update = proto.hasUpdate() ? new Update(proto.getUpdate()) : null;
    }

    public static AccountInformation generateAccountInformation(Starcoin.ProtoAccountInfo proto) {
        if (proto == null) {
            return null;
        }
        AccountInformation accountInformation = new AccountInformation();
        accountInformation.unmarshalProto(proto);
        return accountInformation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof AccountInformation)) {
            return false;
        }
        AccountInformation that = (AccountInformation) o;
        return allotment == that.allotment &&
                Objects.equals(address, that.address) &&
                Objects.equals(update, that.update);
    }

    @Override
    public int hashCode() {
        return Objects.hash(address, allotment, update);
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/AccountInformation.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.apache.commons.lang3.RandomUtils
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema

//TODO rename to AccountInfo
@Serializable
@ProtobufSchema(Starcoin.ProtoAccountInfo::class)
data class AccountInformation(@SerialId(1) val addressHash: Hash, @SerialId(2) var update: Update, @SerialId(3) var allotment: Long = 0) :
    SiriusObject() {

    constructor(
        address: Address,
        update: Update,
        allotment: Long
    ) : this(address.hash(), update, allotment)

    companion object : SiriusObjectCompanion<AccountInformation, Starcoin.ProtoAccountInfo>(AccountInformation::class) {
        //TODO ensure empty account
        val EMPTY_ACCOUNT = AccountInformation(Address.DUMMY_ADDRESS, Update(UpdateData()), 0)

        @Deprecated("Please use parseFromProtoMessage")
        fun generateAccountInformation(proto: Starcoin.ProtoAccountInfo): AccountInformation? {
            return parseFromProtoMessage(proto)
        }

        override fun mock(): AccountInformation {
            return AccountInformation(Hash.random(), Update.mock(), RandomUtils.nextLong())
        }

        override fun parseFromProtoMessage(protoMessage: Starcoin.ProtoAccountInfo): AccountInformation {
            return AccountInformation(
                Hash.wrap(protoMessage.addressHash),
                Update.parseFromProtoMessage(protoMessage.update),
                protoMessage.allotment
            )
        }

        override fun toProtoMessage(obj: AccountInformation): Starcoin.ProtoAccountInfo {
            return Starcoin.ProtoAccountInfo.newBuilder().setAddressHash(obj.addressHash.toByteString())
                .setUpdate(Update.toProtoMessage(obj.update)).setAllotment(obj.allotment).build()
        }
    }
}
